metrics/src/metrics.rs

```diff
@@ -189,60 +242,40 @@ impl MetricsAgent {
                 }
             }
 
-            let now = Instant::now();
-            if now.duration_since(last_write_time) >= write_frequency_secs && !points.is_empty() {
-                let num_points = points.len();
-                let points_written;
-                debug!("run: attempting to write {} points", points.len());
-                if points.len() > max_points {
-                    warn!(
-                        "max submission rate of {} datapoints per second exceeded.  only the
-                    first {} of {} points will be submitted",
-                        max_points_per_sec,
-                        max_points,
-                        points.len()
-                    );
-                    points.truncate(max_points - 1);
-                }
-                points_written = points.len();
-
-                points.push(
-                    influxdb::Point::new("metrics")
-                        .add_timestamp(timing::timestamp() as i64)
-                        .add_field("host_id", influxdb::Value::String(HOST_INFO.to_string()))
-                        .add_field(
-                            "points_written",
-                            influxdb::Value::Integer(points_written as i64),
-                        )
-                        .add_field("num_points", influxdb::Value::Integer(num_points as i64))
-                        .add_field(
-                            "secs_since_last_write",
-                            influxdb::Value::Integer(
-                                now.duration_since(last_write_time).as_secs() as i64
-                            ),
-                        )
-                        .add_field(
-                            "points_rate_exceeded",
-                            influxdb::Value::Boolean(num_points > max_points),
-                        )
-                        .to_owned(),
+            let mut num_max_writes = max_points;
+
+            points_map.retain(|_level, (last_time, points)| {
```

Comment:
I think this needs a bit of work, as the hashmap is not sorted based on keys (log levels). So it's not rate limiting in the right order. I'll push another patch in this PR to fix it.